### PR TITLE
p256: add 32-bit arithmetic and optimizations

### PR DESCRIPTION
**Pull request by davideciacciolo**
This patch adds 32-bit proper arithmetic to scalar and field operations. Certain algorithms (like barret reduction, and, most of all, montgomery reduction) offer certain optimization when the word size is 32-bit long instead of 64.

### CODE REVIEW

**Comment by tarcieri** [comment]
It'd be good to see benchmarks for this. See #885 where we reverted previous attempts to add a 32-bit backend due to performance regressions.

**Comment by davideciacciolo** [comment]
The improvement I registered is for ECDSA signature verification.

**Comment by tarcieri** [comment]
There are existing benchmarks you can run with `cargo bench`

**Comment by davideciacciolo** [comment]
Ok, this is the result I got with the modified version:
```
point operations/point-scalar mul
                        time:   [111.31 µs 111.53 µs 111.81 µs]
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) high mild
  6 (6.00%) high severe

scalar operations/sub   time:   [3.0671 ns 3.0752 ns 3.0844 ns]
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe
scalar operations/add   time:   [4.9285 ns 4.9330 ns 4.9384 ns]
Found 11 outliers among 100 measurements (11.00%)
  7 (7.00%) high mild
  4 (4.00%) high severe
scalar operations/mul   time:   [42.706 ns 42.730 ns 42.756 ns]
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe
scalar operations/negate
                        time:   [2.7322 ns 2.7347 ns 2.7378 ns]
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) high mild
  9 (9.00%) high severe
scalar operations/invert
                        time:   [17.088 µs 17.096 µs 17.105 µs]
Found 9 outliers among 100 measurements (9.00%)
  2 (2.00%) high mild
  7 (7.00%) high severe
```

**Comment by tarcieri** [comment]
Can you run it once on `master`, and then run it with your changes? It will show the difference.

**Comment by davideciacciolo** [comment]
Yes, I was about to send it. This is the non-modified version:
```
point operations/point-scalar mul
                        time:   [107.60 µs 107.67 µs 107.76 µs]
Found 15 outliers among 100 measurements (15.00%)
  4 (4.00%) high mild
  11 (11.00%) high severe

scalar operations/sub   time:   [3.0710 ns 3.0760 ns 3.0818 ns]
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe
scalar operations/add   time:   [4.9870 ns 5.0166 ns 5.0525 ns]
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
scalar operations/mul   time:   [47.291 ns 47.758 ns 48.218 ns]
scalar operations/negate
                        time:   [2.7554 ns 2.7586 ns 2.7621 ns]
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) low mild
  4 (4.00%) high mild
  2 (2.00%) high severe
scalar operations/invert
                        time:   [17.306 µs 17.394 µs 17.509 µs]
Found 14 outliers among 100 measurements (14.00%)
  8 (8.00%) high mild
  6 (6.00%) high severe
```

**Comment by tarcieri** [comment]
Okay, so not as bad as #882, but still seems like a slight regression

**Comment by davideciacciolo** [comment]
I tried it again and it shows different results, I sincerely don't know since it's the first time I use `cargo bench`:
```
point operations/point-scalar mul
                        time:   [108.69 µs 108.85 µs 109.03 µs]
                        change: [-3.9125% -2.1764% -0.2705%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 17 outliers among 100 measurements (17.00%)
  5 (5.00%) high mild
  12 (12.00%) high severe

scalar operations/sub   time:   [3.0630 ns 3.0686 ns 3.0752 ns]
                        change: [-1.1825% -0.2047% +1.0201%] (p = 0.75 > 0.05)
                        No change in performance detected.
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe
scalar operations/add   time:   [4.7953 ns 4.8105 ns 4.8265 ns]
                        change: [-4.4281% -3.0878% -1.9844%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe
scalar operations/mul   time:   [41.503 ns 41.541 ns 41.580 ns]
                        change: [-3.9748% -2.6948% -1.4332%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) high mild
  9 (9.00%) high severe
scalar operations/negate
                        time:   [2.7340 ns 2.7359 ns 2.7379 ns]
                        change: [-1.2170% -0.0645% +1.1043%] (p = 0.92 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  3 (3.00%) high mild
  6 (6.00%) high severe
scalar operations/invert
                        time:   [17.117 µs 17.130 µs 17.145 µs]
                        change: [-1.5923% -0.2443% +0.6215%] (p = 0.78 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  8 (8.00%) high severe
  ```
(This is the modified version)

**Comment by tarcieri** [comment]
Aah, perhaps it's just noise then.

I can pull it down locally and check.

File: p256/src/arithmetic/field/field32.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+use super::MODULUS;
+use crate::arithmetic::util::*;
+use elliptic_curve::bigint::{U256, U512};
+
+pub(super) const fn add(a: U256, b: U256) -> U256 {
+    let a = a.as_words();
+    let b = b.as_words();
+
+    // Bit 256 of p is set, so addition can result in nine words.
+    let (w0, carry) = adc(a[0], b[0], 0);
```
Comment (tarcieri) [inline p256/src/arithmetic/field/field32.rs:12]: Note that if you used `as_limbs`, they define a word size specific `Limb::adc` operation which could be used here instead (and get rid of `util.rs` in the process), however that's a refactoring I can potentially do as a followup.

**Comment by davideciacciolo** [comment]
Refactoring done. Regarding benchmarks, 64-bit code didn't change almost at all, so I think too it should be noise. I don't know if there is a way to benchmark in a systematic way on a separate device using `cargo bench` (for 32-bit systems). I'm quite sure there is a good improvement since for ECDSA verification it was 4-5% faster, but I didn't test the `p256` crate standalone.

**Review by tarcieri — APPROVED** [review]

**Comment by tarcieri** [comment]
Thank you!

**Pull request merged**